ckan/types.py

```diff
@@ -0,0 +1,271 @@
+    resource: "model_.Resource"
+    resource_view: "model_.ResourceView"
+    relationship: "model_.PackageRelationship"
+    api_version: int
+    dataset_counts: Dict[str, Any]
+    limits: Dict[str, Any]
+    metadata_modified: str
+    with_capacity: bool
+
+    table_names: List[str]
+
+
+class AuthResult(TypedDict, total=False):
+    success: bool
+    msg: Optional[str]
+
+
+class ValueValidator(Protocol):
+    def __call__(self, value: Any) -> Any:
+        ...
+
+
+class ContextValidator(Protocol):
+    def __call__(self, value: Any, context: Context) -> Any:
+        ...
+
+
+class DataValidator(Protocol):
+    def __call__(
+        self,
+        key: FlattenKey,
+        data: Dict[FlattenKey, Any],
+        errors: FlattenErrorDict,
+        context: Context,
+    ) -> None:
+        ...
+
+
+Validator = Union[ValueValidator, ContextValidator, DataValidator]
+
+NestedSchema = Dict[str, Iterable[Validator]]
+# Schema = Dict[str, Union[Iterable[Validator], "NestedSchema"]]
+Schema = Dict[str, Union[Iterable[Validator], "Schema"]]
+ComplexSchemaFunc = Callable[..., Schema]
+PlainSchemaFunc = Callable[[], Schema]
+
+AuthFunctionWithOptionalDataDict = Callable[
+    [Context, Optional[DataDict]], AuthResult
+]
+AuthFunctionWithMandatoryDataDict = Callable[[Context, DataDict], AuthResult]
+AuthFunction = Union[
+    AuthFunctionWithOptionalDataDict,
+    AuthFunctionWithMandatoryDataDict,
+    'partial[AuthResult]',
+]
+ChainedAuthFunction = Callable[
+    [AuthFunction, Context, Optional[DataDict]], AuthResult
+]
+
+Action = Callable[[Context, DataDict], Any]
+ChainedAction = Callable[[Action, Context, DataDict], Any]
+
+
+class PFeed(Protocol):
+    def __init__(
+        self,
+        feed_title: str,
+        feed_link: str,
+        feed_description: str,
+        language: Optional[str],
+        author_name: Optional[str],
+        feed_guid: Optional[str],
+        feed_url: Optional[str],
+        previous_page: Optional[str],
+        next_page: Optional[str],
+        first_page: Optional[str],
+        last_page: Optional[str],
+    ) -> None:
+        ...
+
+    def add_item(self, **kwargs: Any) -> None:
+        ...
+
+    def writeString(self, encoding: str) -> str:
+        ...
+
+
+class PUploader(Protocol):
+    def __init__(
+        self, object_type: str, old_filename: Optional[str] = None
+    ) -> None:
+        ...
+
+    def upload(self, max_size: int = ...) -> None:
+        ...
+
+    def update_data_dict(
+        self,
+        data_dict: Dict[str, Any],
+        url_field: str,
+        file_field: str,
+        clear_field: str,
+    ) -> None:
+        ...
+
+
+class PResourceUploader(Protocol):
+    mimetype: Optional[str]
+    filesize: int
+
+    def __init__(self, resource: Dict[str, Any]) -> None:
+        ...
+
+    def get_path(self, id: str) -> str:
+        ...
+
+    def upload(self, id: str, max_size: int = ...) -> None:
+        ...
+
+
+class PModel(Protocol):
```

Comment:
@smotornyuk would you add an explanation of what this is for, i.e. "may be useful if we start using a type checker other than pyright"?

Reply:
For now, I've removed it. If someone will be interested in it, I always can point a person to old git commit